python/packages/autogen-core/docs/src/user-guide/core-user-guide/cookbook/distributed-group-chat.ipynb

```diff
@@ -0,0 +1,258 @@
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class ParticipantAgent(RoutedAgent):\n",
+    "    def __init__(self, name: str, description: str, model_client: ChatCompletionClient, system_message: str) -> None:\n",
+    "        super().__init__(\"A Participant\")\n",
+    "        self._name = name\n",
+    "        self._description = description\n",
+    "        self._model_client = model_client\n",
+    "        self._chat_history: List[LLMMessage] = [SystemMessage(system_message)]\n",
+    "\n",
+    "    @message_handler\n",
+    "    async def handle_message(self, message: GroupChatMessage, ctx: MessageContext) -> None:\n",
+    "        print(f\"received {message}\")\n",
+    "        self._chat_history.append(message.body)\n",
+    "\n",
+    "    @message_handler\n",
+    "    async def handle_request_to_speak(self, message: RequestToSpeak, ctx: MessageContext) -> None:\n",
+    "        completion = await self._model_client.create(self._chat_history)\n",
+    "        assert isinstance(completion.content, str)\n",
+    "        self._chat_history.append(AssistantMessage(content=completion.content, source=self._name))\n",
+    "        print(f\"\\n{'-'*80}\\n{self._name} ({id(self)}):\\n{completion.content}\")\n",
+    "        await asyncio.sleep(2)\n",
+    "        await self.publish_message(\n",
+    "            GroupChatMessage(body=UserMessage(content=completion.content, source=self._name)),\n",
+    "            TopicId(type=\"agent_result\", source=\"participant\"),\n",
+    "        )\n",
+    "\n",
+    "\n",
+    "@type_subscription(topic_type=\"agent_result\")\n",
+    "class GroupChatManager(RoutedAgent):\n",
+    "    def __init__(self, participant_topic_types: List[str], max_rounds: int = 3) -> None:\n",
+    "        super().__init__(\"Group chat manager\")\n",
+    "        self._num_rounds = 1\n",
+    "        self._participant_topic_types = participant_topic_types\n",
+    "        self._chat_history: List[GroupChatMessage] = []\n",
+    "        self._max_rounds = max_rounds\n",
+    "\n",
+    "    @message_handler\n",
+    "    async def handle_message(self, message: GroupChatMessage, ctx: MessageContext) -> None:\n",
+    "        self._chat_history.append(message)\n",
+    "        assert isinstance(message.body, UserMessage)\n",
+    "        if self._num_rounds >= self._max_rounds:\n",
+    "            print(f\"\\n\\n\\n ---> Finished running {self._num_rounds} rounds! <---\")\n",
+    "            await asyncio.sleep(1)\n",
+    "            return\n",
+    "        speaker_topic_type = self._participant_topic_types[self._num_rounds % len(self._participant_topic_types)]\n",
+    "        self._num_rounds += 1\n",
+    "        print(f\"\\n{'-'*80}\\n Manager ({id(self)}): Asking {speaker_topic_type} to speak\")\n",
+    "        await self.publish_message(RequestToSpeak(), TopicId(type=speaker_topic_type, source=\"manager\"))"
```

Comment:
Same, don't specify the source of topic, the source should not be hard-coded inside agent.